File: back/microservices/communications/ms-calls/src/stores/calls/webm.ts
/**
 * Pure-TS EBML/WebM muxer for Opus audio.
 *
 * Packs a list of raw Opus frames (each one a 20 ms packet, as captured from
 * the WebRTC RTP stream by centimanus) into a valid WebM/Opus file that
 * browsers play natively and Web Audio `decodeAudioData` can decode.
 *
 * Timecodes are laid out SEQUENTIALLY at 20 ms per frame — exactly like the
 * original Go recorder (`int64(i * 20)`). Wall-clock arrival timestamps carry
 * network/scheduler jitter and, when used as block timecodes, make playback
 * choppy/garbled ("каша"). Each Opus frame is a fixed 20 ms slice, so a clean
 * sequential timeline reconstructs the audio faithfully.
 */

// EBML element IDs (full IDs with marker bits, as stored on the wire).
const ID_EBML = 0x1a45dfa3;
const ID_SEGMENT = 0x18538067;
const ID_INFO = 0x1549a966;
const ID_TIMESTAMP_SCALE = 0x2ad7b1;
const ID_MUXING_APP = 0x4d80;
const ID_WRITING_APP = 0x5741;
const ID_DURATION = 0x4489;
const ID_TRACKS = 0x1654ae6b;
const ID_TRACK_ENTRY = 0xae;
const ID_TRACK_NUMBER = 0xd7;
const ID_TRACK_UID = 0x73c5;
const ID_TRACK_TYPE = 0x83;
const ID_CODEC_ID = 0x86;
const ID_CODEC_PRIVATE = 0x63a2;
const ID_AUDIO = 0xe1;
const ID_SAMPLING_FREQ = 0xb5;
const ID_CHANNELS = 0x9f;
const ID_CLUSTER = 0x1f43b675;
const ID_CLUSTER_TIMESTAMP = 0xe7;
const ID_SIMPLE_BLOCK = 0xa3;

const ID_EBML_VERSION = 0x4286;
const ID_EBML_READ_VERSION = 0x42f7;
const ID_EBML_MAX_ID_LEN = 0x42f2;
const ID_EBML_MAX_SIZE_LEN = 0x42f3;
const ID_DOCTYPE = 0x4282;
const ID_DOCTYPE_VERSION = 0x4287;
const ID_DOCTYPE_READ_VERSION = 0x4285;

/** Each Opus frame is a fixed 20 ms slice. */
const FRAME_MS = 20;
/** Start a new cluster at least every ~30 s of audio. */
const CLUSTER_MAX_MS = 30_000;
const FRAMES_PER_CLUSTER = Math.floor(CLUSTER_MAX_MS / FRAME_MS);

export type WebMOpusOptions = {
  sampleRate?: number;
  channels?: number;
};

/** Encode an EBML element ID into its minimal big-endian byte form. */
function encodeId(id: number): Buffer {
  if (id <= 0xff) return Buffer.from([id]);
  if (id <= 0xffff) return Buffer.from([id >> 8, id & 0xff]);
  if (id <= 0xffffff) return Buffer.from([id >> 16, (id >> 8) & 0xff, id & 0xff]);
  return Buffer.from([id >>> 24, (id >> 16) & 0xff, (id >> 8) & 0xff, id & 0xff]);
}

/** Encode a length as an EBML variable-size integer (VINT). */
function encodeVintSize(size: number): Buffer {
  if (size < 0x7f) return Buffer.from([0x80 | size]);
  if (size < 0x3ffe) return Buffer.from([0x40 | (size >> 8), size & 0xff]);
  if (size < 0x1ffffe)
    return Buffer.from([0x20 | (size >> 16), (size >> 8) & 0xff, size & 0xff]);
  if (size < 0x0ffffffe)
    return Buffer.from([
      0x10 | (size >> 24),
      (size >> 16) & 0xff,
      (size >> 8) & 0xff,
      size & 0xff,
    ]);
  // 8-byte VINT for very large elements.
  const out = Buffer.alloc(8);
  out[0] = 0x01;
  for (let i = 1; i < 8; i++) {
    out[i] = (size / 2 ** (8 * (7 - i))) & 0xff;
  }
  return out;
}

/** A full element: id + VINT(size) + body. */
function elem(id: number, body: Buffer): Buffer {
  return Buffer.concat([encodeId(id), encodeVintSize(body.length), body]);
}

function uintElem(id: number, value: number, size: number): Buffer {
  const body = Buffer.alloc(size);
  for (let i = 0; i < size; i++) {
    body[size - 1 - i] = (value / 2 ** (8 * i)) & 0xff;
  }
  return elem(id, body);
}

function f64Elem(id: number, value: number): Buffer {
  const body = Buffer.alloc(8);
  body.writeDoubleBE(value, 0);
  return elem(id, body);
}

function strElem(id: number, s: string): Buffer {
  return elem(id, Buffer.from(s, "utf8"));
}

/** 19-byte OpusHead identification header for CodecPrivate. */
function buildOpusHead(sampleRate: number, channels: number): Buffer {
  const h = Buffer.alloc(19);
  h.write("OpusHead", 0, "ascii");
  h[8] = 1; // version
  h[9] = channels; // channel count
  h.writeUInt16LE(0, 10); // pre-skip
  h.writeUInt32LE(sampleRate, 12); // input sample rate
  h.writeInt16LE(0, 16); // output gain
  h[18] = 0; // channel mapping family
  return h;
}

function buildEbmlHeader(): Buffer {
  const body = Buffer.concat([
    uintElem(ID_EBML_VERSION, 1, 1),
    uintElem(ID_EBML_READ_VERSION, 1, 1),
    uintElem(ID_EBML_MAX_ID_LEN, 4, 1),
    uintElem(ID_EBML_MAX_SIZE_LEN, 8, 1),
    strElem(ID_DOCTYPE, "webm"),
    uintElem(ID_DOCTYPE_VERSION, 4, 1),
    uintElem(ID_DOCTYPE_READ_VERSION, 2, 1),
  ]);
  return elem(ID_EBML, body);
}

function buildInfo(frameCount: number): Buffer {
  const durationMs = frameCount * FRAME_MS;
  const body = Buffer.concat([
    uintElem(ID_TIMESTAMP_SCALE, 1_000_000, 4), // 1 ms ticks
    strElem(ID_MUXING_APP, "ms-calls"),
    strElem(ID_WRITING_APP, "ms-calls"),
    f64Elem(ID_DURATION, durationMs),
  ]);
  return elem(ID_INFO, body);
}

function buildTracks(sampleRate: number, channels: number): Buffer {
  const audio = Buffer.concat([
    f64Elem(ID_SAMPLING_FREQ, sampleRate),
    uintElem(ID_CHANNELS, channels, 1),
  ]);
  const track = Buffer.concat([
    uintElem(ID_TRACK_NUMBER, 1, 1),
    uintElem(ID_TRACK_UID, 1, 1),
    uintElem(ID_TRACK_TYPE, 2, 1), // audio
    strElem(ID_CODEC_ID, "A_OPUS"),
    elem(ID_CODEC_PRIVATE, buildOpusHead(sampleRate, channels)),
    elem(ID_AUDIO, audio),
  ]);
  return elem(ID_TRACKS, elem(ID_TRACK_ENTRY, track));
}

function buildSimpleBlock(relMs: number, frame: Uint8Array): Buffer {
  const header = Buffer.from([
    0x81, // track number VINT = 1
    (relMs >> 8) & 0xff,
    relMs & 0xff,
    0x80, // keyframe flag
  ]);
  return elem(ID_SIMPLE_BLOCK, Buffer.concat([header, Buffer.from(frame)]));
}

function buildClusters(frames: Uint8Array[]): Buffer {
  const clusters: Buffer[] = [];
  for (let start = 0; start < frames.length; start += FRAMES_PER_CLUSTER) {
    const end = Math.min(start + FRAMES_PER_CLUSTER, frames.length);
    const baseMs = start * FRAME_MS;

    const blocks: Buffer[] = [uintElem(ID_CLUSTER_TIMESTAMP, baseMs, 4)];
    for (let i = start; i < end; i++) {
      const relMs = i * FRAME_MS - baseMs; // < CLUSTER_MAX_MS, fits in i16
      blocks.push(buildSimpleBlock(relMs, frames[i]));
    }
    clusters.push(elem(ID_CLUSTER, Buffer.concat(blocks)));
  }
  return Buffer.concat(clusters);
}

/**
 * Build a complete WebM/Opus file from sequential 20 ms Opus frames.
 * Returns `undefined` when there are no frames.
 */
export function writeWebMOpus(
  frames: Uint8Array[],
  opts: WebMOpusOptions = {},
): Uint8Array | undefined {
  if (frames.length === 0) return undefined;
  const sampleRate = opts.sampleRate ?? 48000;
  const channels = opts.channels ?? 1;

  // Segment with unknown size (0x01 + seven 0xFF), so we can stream clusters.
  const segmentUnknownSize = Buffer.from([
    0x01, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
  ]);

  return Buffer.concat([
    buildEbmlHeader(),
    encodeId(ID_SEGMENT),
    segmentUnknownSize,
    buildInfo(frames.length),
    buildTracks(sampleRate, channels),
    buildClusters(frames),
  ]);
}

File: native/resonus/src/webm/webm.zig
// Pure-Zig EBML/WebM writer for Opus audio.
// Produces a valid WebM file playable in browsers from a slice of Opus frames.
const std = @import("std");

// EBML element IDs (pre-encoded with marker bits)
const ID_EBML: u32 = 0x1A45DFA3;
const ID_SEGMENT: u32 = 0x18538067;
const ID_INFO: u32 = 0x1549A966;
const ID_TIMESTAMP_SCALE: u32 = 0x2AD7B1;
const ID_MUXING_APP: u32 = 0x4D80;
const ID_WRITING_APP: u32 = 0x5741;
const ID_DURATION: u32 = 0x4489;
const ID_TRACKS: u32 = 0x1654AE6B;
const ID_TRACK_ENTRY: u32 = 0xAE;
const ID_TRACK_NUMBER: u32 = 0xD7;
const ID_TRACK_UID: u32 = 0x73C5;
const ID_TRACK_TYPE: u32 = 0x83;
const ID_CODEC_ID: u32 = 0x86;
const ID_CODEC_PRIVATE: u32 = 0x63A2;
const ID_AUDIO: u32 = 0xE1;
const ID_SAMPLING_FREQ: u32 = 0xB5;
const ID_CHANNELS: u32 = 0x9F;
const ID_CLUSTER: u32 = 0x1F43B675;
const ID_CLUSTER_TIMESTAMP: u32 = 0xE7;
const ID_SIMPLE_BLOCK: u32 = 0xA3;

// WebM EBML element IDs
const ID_EBML_VERSION: u32 = 0x4286;
const ID_EBML_READ_VERSION: u32 = 0x42F7;
const ID_EBML_MAX_ID_LEN: u32 = 0x42F2;
const ID_EBML_MAX_SIZE_LEN: u32 = 0x42F3;
const ID_DOCTYPE: u32 = 0x4282;
const ID_DOCTYPE_VERSION: u32 = 0x4287;
const ID_DOCTYPE_READ_VERSION: u32 = 0x4285;

const CLUSTER_MAX_MS: u32 = 30_000;

pub const OpusFrame = struct {
    timestamp_ms: u32,
    data: []const u8,
};

/// Writes a complete WebM file and returns the bytes (caller frees).
pub fn writeWebM(allocator: std.mem.Allocator, frames: []const OpusFrame, sample_rate: u32, channels: u8) ![]u8 {
    if (frames.len == 0) return error.NoFrames;

    var out = try std.ArrayList(u8).initCapacity(allocator, 4096);
    defer out.deinit(allocator);

    try buildEbmlHeader(&out, allocator);
    try buildSegment(&out, allocator, frames, sample_rate, channels);

    return out.toOwnedSlice(allocator);
}

fn buildEbmlHeader(out: *std.ArrayList(u8), allocator: std.mem.Allocator) !void {
    var body = try std.ArrayList(u8).initCapacity(allocator, 64);
    defer body.deinit(allocator);

    try writeUintElem(&body, allocator, ID_EBML_VERSION, 1, 1);
    try writeUintElem(&body, allocator, ID_EBML_READ_VERSION, 1, 1);
    try writeUintElem(&body, allocator, ID_EBML_MAX_ID_LEN, 4, 1);
    try writeUintElem(&body, allocator, ID_EBML_MAX_SIZE_LEN, 8, 1);
    try writeStrElem(&body, allocator, ID_DOCTYPE, "webm");
    try writeUintElem(&body, allocator, ID_DOCTYPE_VERSION, 4, 1);
    try writeUintElem(&body, allocator, ID_DOCTYPE_READ_VERSION, 2, 1);

    const bytes = try body.toOwnedSlice(allocator);
    defer allocator.free(bytes);
    try writeElem(out, allocator, ID_EBML, bytes);
}

fn buildSegment(out: *std.ArrayList(u8), allocator: std.mem.Allocator, frames: []const OpusFrame, sample_rate: u32, channels: u8) !void {
    // Segment with unknown size
    try writeId(out, allocator, ID_SEGMENT);
    try out.appendSlice(allocator, &[_]u8{ 0x01, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF });

    try buildInfo(out, allocator, frames);
    try buildTracks(out, allocator, sample_rate, channels);
    try buildClusters(out, allocator, frames);
}

fn buildInfo(out: *std.ArrayList(u8), allocator: std.mem.Allocator, frames: []const OpusFrame) !void {
    var body = try std.ArrayList(u8).initCapacity(allocator, 64);
    defer body.deinit(allocator);

    try writeUintElem(&body, allocator, ID_TIMESTAMP_SCALE, 1_000_000, 4);
    try writeStrElem(&body, allocator, ID_MUXING_APP, "resonus");
    try writeStrElem(&body, allocator, ID_WRITING_APP, "resonus");

    const last_ms = frames[frames.len - 1].timestamp_ms;
    const duration_ms: f64 = @floatFromInt(last_ms + 20);
    try writeF64Elem(&body, allocator, ID_DURATION, duration_ms);

    const bytes = try body.toOwnedSlice(allocator);
    defer allocator.free(bytes);
    try writeElem(out, allocator, ID_INFO, bytes);
}

fn buildTracks(out: *std.ArrayList(u8), allocator: std.mem.Allocator, sample_rate: u32, channels: u8) !void {
    var track = try std.ArrayList(u8).initCapacity(allocator, 128);
    defer track.deinit(allocator);

    try writeUintElem(&track, allocator, ID_TRACK_NUMBER, 1, 1);
    try writeUintElem(&track, allocator, ID_TRACK_UID, 1, 1);
    try writeUintElem(&track, allocator, ID_TRACK_TYPE, 2, 1); // audio
    try writeStrElem(&track, allocator, ID_CODEC_ID, "A_OPUS");

    const opus_head = buildOpusHead(sample_rate, channels);
    try writeElem(&track, allocator, ID_CODEC_PRIVATE, &opus_head);

    var audio = try std.ArrayList(u8).initCapacity(allocator, 16);
    defer audio.deinit(allocator);
    try writeF64Elem(&audio, allocator, ID_SAMPLING_FREQ, @floatFromInt(sample_rate));
    try writeUintElem(&audio, allocator, ID_CHANNELS, channels, 1);
    const audio_bytes = try audio.toOwnedSlice(allocator);
    defer allocator.free(audio_bytes);
    try writeElem(&track, allocator, ID_AUDIO, audio_bytes);

    const track_bytes = try track.toOwnedSlice(allocator);
    defer allocator.free(track_bytes);

    var tracks = try std.ArrayList(u8).initCapacity(allocator, track_bytes.len + 4);
    defer tracks.deinit(allocator);
    try writeElem(&tracks, allocator, ID_TRACK_ENTRY, track_bytes);

    const tracks_bytes = try tracks.toOwnedSlice(allocator);
    defer allocator.free(tracks_bytes);
    try writeElem(out, allocator, ID_TRACKS, tracks_bytes);
}

fn buildClusters(out: *std.ArrayList(u8), allocator: std.mem.Allocator, frames: []const OpusFrame) !void {
    var start: usize = 0;
    while (start < frames.len) {
        const base_ms = frames[start].timestamp_ms;

        var end = start;
        while (end < frames.len and frames[end].timestamp_ms - base_ms < CLUSTER_MAX_MS) {
            end += 1;
        }

        var cluster = try std.ArrayList(u8).initCapacity(allocator, 1024);
        defer cluster.deinit(allocator);

        try writeUintElem(&cluster, allocator, ID_CLUSTER_TIMESTAMP, base_ms, 4);

        for (frames[start..end]) |frame| {
            const rel = @min(frame.timestamp_ms - base_ms, 32767);
            const rel_ms: u16 = @truncate(rel);

            var sb = try std.ArrayList(u8).initCapacity(allocator, 4 + frame.data.len);
            defer sb.deinit(allocator);
            try sb.append(allocator, 0x81); // track VINT = 1
            try sb.append(allocator, @truncate(rel_ms >> 8));
            try sb.append(allocator, @truncate(rel_ms & 0xFF));
            try sb.append(allocator, 0x80); // keyframe flag
            try sb.appendSlice(allocator, frame.data);

            const sb_bytes = try sb.toOwnedSlice(allocator);
            defer allocator.free(sb_bytes);
            try writeElem(&cluster, allocator, ID_SIMPLE_BLOCK, sb_bytes);
        }

        const cluster_bytes = try cluster.toOwnedSlice(allocator);
        defer allocator.free(cluster_bytes);
        try writeElem(out, allocator, ID_CLUSTER, cluster_bytes);

        start = end;
    }
}

fn buildOpusHead(sample_rate: u32, channels: u8) [19]u8 {
    var h: [19]u8 = undefined;
    h[0] = 'O';
    h[1] = 'p';
    h[2] = 'u';
    h[3] = 's';
    h[4] = 'H';
    h[5] = 'e';
    h[6] = 'a';
    h[7] = 'd';
    h[8] = 1; // version
    h[9] = channels; // channel count
    h[10] = 0;
    h[11] = 0; // pre-skip LE u16
    h[12] = @truncate(sample_rate);
    h[13] = @truncate(sample_rate >> 8);
    h[14] = @truncate(sample_rate >> 16);
    h[15] = @truncate(sample_rate >> 24);
    h[16] = 0;
    h[17] = 0; // output gain LE i16
    h[18] = 0; // channel mapping family
    return h;
}

fn writeElem(buf: *std.ArrayList(u8), allocator: std.mem.Allocator, id: u32, body: []const u8) !void {
    try writeId(buf, allocator, id);
    try writeVintSize(buf, allocator, body.len);
    try buf.appendSlice(allocator, body);
}

fn writeUintElem(buf: *std.ArrayList(u8), allocator: std.mem.Allocator, id: u32, value: u64, size: u4) !void {
    var body: [8]u8 = undefined;
    var i: u4 = 0;
    while (i < size) : (i += 1) {
        body[size - 1 - i] = @truncate((value >> (@as(u6, i) * 8)) & 0xFF);
    }
    try writeElem(buf, allocator, id, body[0..size]);
}

fn writeF64Elem(buf: *std.ArrayList(u8), allocator: std.mem.Allocator, id: u32, value: f64) !void {
    const bits: u64 = @bitCast(value);
    var body: [8]u8 = undefined;
    inline for (0..8) |i| {
        body[i] = @truncate(bits >> (56 - @as(u6, i) * 8));
    }
    try writeElem(buf, allocator, id, &body);
}

fn writeStrElem(buf: *std.ArrayList(u8), allocator: std.mem.Allocator, id: u32, s: []const u8) !void {
    try writeElem(buf, allocator, id, s);
}

fn writeId(buf: *std.ArrayList(u8), allocator: std.mem.Allocator, id: u32) !void {
    if (id <= 0xFF) {
        try buf.append(allocator, @truncate(id));
    } else if (id <= 0xFFFF) {
        try buf.append(allocator, @truncate(id >> 8));
        try buf.append(allocator, @truncate(id & 0xFF));
    } else if (id <= 0xFFFFFF) {
        try buf.append(allocator, @truncate(id >> 16));
        try buf.append(allocator, @truncate((id >> 8) & 0xFF));
        try buf.append(allocator, @truncate(id & 0xFF));
    } else {
        try buf.append(allocator, @truncate(id >> 24));
        try buf.append(allocator, @truncate((id >> 16) & 0xFF));
        try buf.append(allocator, @truncate((id >> 8) & 0xFF));
        try buf.append(allocator, @truncate(id & 0xFF));
    }
}

fn writeVintSize(buf: *std.ArrayList(u8), allocator: std.mem.Allocator, size: usize) !void {
    if (size < 0x7F) {
        try buf.append(allocator, @truncate(0x80 | size));
    } else if (size < 0x3FFE) {
        try buf.append(allocator, @truncate(0x40 | (size >> 8)));
        try buf.append(allocator, @truncate(size & 0xFF));
    } else if (size < 0x1FFFFE) {
        try buf.append(allocator, @truncate(0x20 | (size >> 16)));
        try buf.append(allocator, @truncate((size >> 8) & 0xFF));
        try buf.append(allocator, @truncate(size & 0xFF));
    } else if (size < 0x0FFFFFFE) {
        try buf.append(allocator, @truncate(0x10 | (size >> 24)));
        try buf.append(allocator, @truncate((size >> 16) & 0xFF));
        try buf.append(allocator, @truncate((size >> 8) & 0xFF));
        try buf.append(allocator, @truncate(size & 0xFF));
    } else {
        // 8-byte VINT for very large elements
        try buf.append(allocator, 0x01);
        var shift: i7 = 56;
        while (shift >= 0) : (shift -= 8) {
            try buf.append(allocator, @truncate((size >> @intCast(shift)) & 0xFF));
        }
    }
}
